### ISSUE TITLE
Choosing top-k in n-gram/RNNLM shallow fusion

### ISSUE THREAD

**Issue by huangruizhe** [k2-fsa/icefall#754]
We have seen nice improvement for RNNLM shallow fusion and LODR in this PR #645. 

There might be a small catch in the implementation I hope to discuss, as follows.
https://github.com/k2-fsa/icefall/blob/a0cf85343dad31a678ddaac7652f0bb2bbb4cac2/egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py#L2026

In the above code, we choose the top-k tokens to expand, according to the probability in `ragged_log_probs[i]`.
Then, the score of the extended hypotheses with the chosen tokens are interpolated with an n-gram/RNNLM score:
https://github.com/k2-fsa/icefall/blob/a0cf85343dad31a678ddaac7652f0bb2bbb4cac2/egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py#L2049

The questions is, it seems we should combine the scores **_before_** choosing the top-k, instead of after.
That is, we should probably interpolate the n-gram/RNNLM score with the RNN-T score _**first**_, and then take the top-k tokens from the distribution. This does not bring extra computational cost, at least in the RNNLM case.

What do you think?

**Comment by csukuangfj** [comment]
Here is the logic:

(1) Get the total scores of each hypothesis

https://github.com/k2-fsa/icefall/blob/a0cf85343dad31a678ddaac7652f0bb2bbb4cac2/egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py#L1943-L1945

`ys_log_probs` contains the total scores (including both acoustic scores and LM scores of each hypothesis)

(2) Before expanding, compute the acoustic scores of the current frame
https://github.com/k2-fsa/icefall/blob/a0cf85343dad31a678ddaac7652f0bb2bbb4cac2/egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py#L1970

(3) Add the newly computed acoustic scores to the total scores of each hypothesis so that we can compute the top-k
https://github.com/k2-fsa/icefall/blob/a0cf85343dad31a678ddaac7652f0bb2bbb4cac2/egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py#L1972

(4) Compute top-k
https://github.com/k2-fsa/icefall/blob/a0cf85343dad31a678ddaac7652f0bb2bbb4cac2/egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py#L1996

(5) For each selected hypothesis from (4), expand it and update the total scores with the newly computed LM scores
https://github.com/k2-fsa/icefall/blob/a0cf85343dad31a678ddaac7652f0bb2bbb4cac2/egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py#L2049

---

Regarding your questions:

> The questions is, it seems we should combine the scores before choosing the top-k, instead of after.

If we were doing like this, we would expand the search space significantly before we can choose the top-k.
(That is, for each hypothesis, we need to the compute the LM scores of all possible expansions).

**Comment by huangruizhe** [comment]
Thanks for the very detailed explanation! I agree with you about the logic. This is how it is currently implemented.
I still kinda insist that this will not increase the search space (or please correct me if I am wrong). 

In RNN-T beam search, for each running hypothesis `y` in the beam (`y` is the token sequence without the blank symbol `∅`), we will compute `p(k | t, y)` which is the distribution over the extended vocabulary. This is the RNN-T score.

On the other hand, given the same `y`, we have the RNNLM score `p(k | y)` over the token vocabulary (-- note `k` will not take `∅` here).
I think this has been implemented here:
https://github.com/k2-fsa/icefall/blob/02c18ba4b25a805db8e8dbb6b7fc4766ad1e006a/icefall/rnn_lm/model.py#L171-L173
The logits/softmax should be the distribution over the whole token vocabulary -- making  `p(k | y)` .

To do shallow fusion, we should do `p(k | t, y) + p(k | y)` before taking top-k tokens for `k`, right?
And this should not add more computational cost.

**Comment by huangruizhe** [comment]
>  for each hypothesis, we need to the compute the LM scores of all possible expansions

For each hypothesis, we have both RNN-T score and RNNLM score for all possible expansions, right?

**Comment by marcoyang1998** [comment]
Sorry for the late replay, I was sick during weekend and just get back to work.

> I agree with interpolating the n-gram/RNNLM score with the RNN-T score first， @marcoyang1998 did you have some experiment results about that.

No, I didn't have experiments on this. I also agree with what @huangruizhe said. I will implement a version for that.
The reason why I arrived at the current implementation is that some other toolkits implement shallow fusion in this way.

**Comment by huangruizhe** [comment]
Hope you get well soon!

I found another issue for shallow fusion. Do you think the n-gram LM and RNNLM scores have considered the "</s>" token when finishing a sentence? Or is it necessary? I am not quite sure about this.

**Comment by marcoyang1998** [comment]
> I found another issue for shallow fusion. Do you think the n-gram LM and RNNLM scores have considered the `</s>` token when finishing a sentence? Or is it necessary? I am not quite sure about this.

No, we don't consider this in our current implementation. Do you think it's necessary @csukuangfj ?

A follow-up on adding the `lm_score` before `topk`(I tested on a streaming LSTM model):

Old shallow fusion: 2.42/6.46
New shallow fusion: 2.41/6.29

Both versions have the best `lm_scale` of 0.3 (with a rough grid search). To make the new version work, I actually have to set the scores for blank to zero in the RNNLM so that blank won't be suppressed after log-linear interpolation. Our RNNLM do have a score for the blank symbol though it never appeared in the training transcripts.

I will test on other models and come back with more results.

**Comment by marcoyang1998** [comment]
I've tested it on another model. This time I got 1.97/4.14 (previously 1.96/4.21) after some tuning.

I think I remember the reason why I didn't implement shallow fusion in this way in the first place. The blank probability in the RNNLM is not well defined so you have to do something special about the blank token. This can be a little ad-hoc and may not work well for all models. I am not fully sure if I will merge the changes.

Regarding the difference between the old and new versions. I think as long as the beam size is reasonably big, the difference will be very small. The tokens having very low acoustic scores will also be less likely to have very large scores in LM.

**Comment by huangruizhe** [comment]
I agree with it. It sounds like a good idea that we can increase the beam size as a workaround/approximation. Thanks a lot for the experiments and clarifications!
